notebooks/04_model_select_and_optimize/nni_surprise_svd.ipynb

```diff
@@ -0,0 +1,835 @@
+    "RATING_METRICS = ['rmse']\n",
+    "RANKING_METRICS = ['precision_at_k', 'ndcg_at_k']  \n",
+    "USERCOL = 'userID'\n",
+    "ITEMCOL = 'itemID'\n",
+    "RECOMMEND_SEEN = False\n",
+    "K = 10\n",
+    "RANDOM_STATE = 0\n",
+    "VERBOSE = True\n",
+    "NUM_EPOCHS = 30\n",
+    "BIASED = True\n",
+    "\n",
+    "script_params = \" \".join([\n",
+    "    '--datastore', DATA_DIR,\n",
+    "    '--train-datapath', TRAIN_FILE_NAME,\n",
+    "    '--validation-datapath', VAL_FILE_NAME,\n",
+    "    '--surprise-reader', 'ml-100k',\n",
+    "    '--rating-metrics', \" \".join(RATING_METRICS),\n",
+    "    '--ranking-metrics', \" \".join(RANKING_METRICS),\n",
+    "    '--usercol', USERCOL,\n",
+    "    '--itemcol', ITEMCOL,\n",
+    "    '--k', str(K),\n",
+    "    '--random-state', str(RANDOM_STATE),\n",
+    "    '--epochs', str(NUM_EPOCHS),\n",
+    "    '--primary-metric', PRIMARY_METRIC\n",
+    "])\n",
+    "\n",
+    "if BIASED:\n",
+    "    script_params += ' --biased'\n",
+    "if VERBOSE:\n",
+    "    script_params += ' --verbose'\n",
+    "if RECOMMEND_SEEN:\n",
+    "    script_params += ' --recommend-seen'"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 53,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# hyperparameters search space\n",
+    "# We do not set 'lr_all' and 'reg_all' because they will be overriden by the other lr_ and reg_ parameters\n",
+    "\n",
+    "hyper_params = {\n",
+    "    'n_factors': {\"_type\": \"choice\", \"_value\": [10, 50, 100, 150, 200]},\n",
+    "    'init_mean': {\"_type\": \"uniform\", \"_value\": [-0.5, 0.5]},\n",
+    "    'init_std_dev': {\"_type\": \"uniform\", \"_value\": [0.01, 0.2]},\n",
+    "    'lr_bu': {\"_type\": \"uniform\", \"_value\": [1e-6, 0.1]}, \n",
+    "    'lr_bi': {\"_type\": \"uniform\", \"_value\": [1e-6, 0.1]}, \n",
+    "    'lr_pu': {\"_type\": \"uniform\", \"_value\": [1e-6, 0.1]}, \n",
+    "    'lr_qi': {\"_type\": \"uniform\", \"_value\": [1e-6, 0.1]}, \n",
+    "    'reg_bu': {\"_type\": \"uniform\", \"_value\": [1e-6, 1]},\n",
+    "    'reg_bi': {\"_type\": \"uniform\", \"_value\": [1e-6, 1]}, \n",
+    "    'reg_pu': {\"_type\": \"uniform\", \"_value\": [1e-6, 1]}, \n",
+    "    'reg_qi': {\"_type\": \"uniform\", \"_value\": [1e-6, 1]}\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 54,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "with open('../../reco_utils/nni/search_space_svd.json', 'w') as fp:\n",
+    "    json.dump(hyper_params, fp)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We also create a yaml file for the configuration of the trials and the tuning algorithm to be used (in this experiment we use TPE). The tuning trials will be executed locally on a [Standard_D16_v3 virtual machine](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/sizes-general#dv3-series-1) (16 vcpus, 64 GB memory)."
```

Comment:
Not sure if there is any information about how NNI uses the underlying multi-core capability in a machine to perform the concurrent run, and how it is compared against a cluster of machines. The reason for the ask is that multi-node may not always be more computationally efficient to multi-core. @scarlett2018

Reply:
I leave the distributed cluster for another PR. It is worth noting that the time required is shorter than the Hyperdrive notebook (which ran a cluster) and NNI seemed to exploit the multiple cores well. I gave it a number of cores comparable to the Hyperdrive cluster, so maybe this favours NNI. I think there are cases when a single machine would be preferable and others when a cluster of nodes would be better, but probably the latter holds with larger data sets than ML100K.